### PR TITLE
Respect PDBs during node upgrades and add test coverage to the ServiceTest upgrade test.

### LINKED ISSUE #38336: GCE Node Upgrade should respect PodDisruptionBudget
As part of [pod safety](https://github.com/kubernetes/kubernetes/pull/34160) for [StatefulSets](https://github.com/kubernetes/features/issues/137), we ensured that [kubectl drain uses the eviction subresource](https://github.com/kubernetes/kubernetes/issues/34277) so it respects [PodDisruptionBudget](https://github.com/kubernetes/features/issues/85).

Unfortunately, there is no script or recommended node upgrade process (AFAIK) that actually uses `kubectl drain`. In particular, the experimental [cluster/gce/upgrade.sh](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/upgrade.sh) script uses GCE [rolling updates](https://github.com/kubernetes/kubernetes/issues/33764) which simply terminate and recreate node VMs.

This issue is to track the need to provide a script or recommended process for upgrading nodes that actually respects PodDisruptionBudgets, so that the guarantees made by things like StatefulSet actually work during upgrades.

cc @kubernetes/sig-apps

### PR DESCRIPTION
**Pull request by mml**
This is still a WIP... needs to be squashed at least, and I don't think it's currently passing until I increase the scale of the RC, but please have a look at the general outline.  Thanks!

Fixes #38336 

@kow3ns @bdbauer @krousey @erictune @maisem @davidopp 

```
On GCE, node upgrades will now respect PodDisruptionBudgets, if present.
```

### CODE REVIEW

**Comment by mml** [comment]
No, actually probably need a release note.

**Comment by mml** [comment]
OK!  This is now passing for me.  Whee!

File: test/e2e/upgrades/services.go
Comment (krousey) [inline test/e2e/upgrades/services.go]: This should not be the default for two reasons. 

  1. The enum could expand to upgrade types that you don't intend to be tested this way.
  2. Some cloud providers might not support a graceful node upgrade like this. There's should be a check for that.

Comment (mml) [inline test/e2e/upgrades/services.go]: Done.

**Comment by mml** [comment]
@krousey PTAL

**Comment by mml** [comment]
@krousey it's squashed now and should have a decent release note.

File: cluster/gce/upgrade.sh
Comment (krousey) [inline cluster/gce/upgrade.sh]: Become young? That's... confusing. Why not wait for recreation to complete?

File: cluster/gce/upgrade.sh
Comment (krousey) [inline cluster/gce/upgrade.sh]: This is very dense to read. I would rather something that checked the recreate operation's status. This can't really tell the difference between a recreate and a system crash/restart. Also, the 60 second window just makes me uncomfortable with timing flakiness. If this comes really quick after starting the cluster, you may mistake the instance as having been restarted. If for some reason this takes too long, then you might miss the window.

What about checking the status of the recreate operation? Something like

```
gcloud compute operations list --limit=1 --filter="targetLink:*${instance} AND operationType:recreate" --format='get(status)'
```

should print "DONE" when it's complete.  It's probably as dense, but avoids timing ambiguities and date math. You shouldcheck with `gcloud operations list --format=json` to make sure I got the filters right.

Comment (krousey) [inline cluster/gce/upgrade.sh]: It's also worth noting that this currently works because kubelet deletes and recreates the node object on upgrade. This would not work if kubelet switched to just updating the node object.

File: test/e2e/framework/service_util.go
Comment (krousey) [inline test/e2e/framework/service_util.go]: ... not your fault, but could use just `var replicas  int32 = 2` at the top of this function and use `&replicas` here.

File: test/e2e/framework/service_util.go
Comment (krousey) [inline test/e2e/framework/service_util.go]: So this is starting to change all of the services e2e tests. I don't know enough about them to say that this is OK or not. You should find the people who own those tests to review this.

**Comment by krousey** [comment]
I only did a quick review earlier. Here's a more in-depth review. Overall it looks good.

**Review by krousey — CHANGES_REQUESTED** [review]

File: cluster/gce/upgrade.sh
Comment (mml) [inline cluster/gce/upgrade.sh]: I agree about the timing flakiness, and I agree I'd like to be sure that the instance has been recreated.

But both of those can happen and it can still be the case that the kubelet hasn't deleted and recreated the node object, so we still need to make sure that the node is `Ready`, and we need to make sure we're checking that the *new* node is healthy, because we may observe the old node with the same name still being `Ready`.

I think I can do this if the script itself deletes the node once it's been drained.  The order would be

1. `kubectl drain`
2. `kubectl delete node`
3. `gcloud ... recreate-instances`
4. poll `gcloud compute operations list`
5. poll `kubectl get node` waiting for it to not have `SchedulingDisabled` and *to* have `Ready`.

This depends on a different assumption about kubelet's behavior, that it won't re-add itself after I delete it.  That seems to be the case if nothing happens on the node, but if kubelet randomly gets restarted (or changes its behavior), it will get re-added, and it will no longer be drained.  IOW, deleting the node opens the door for the node to acquire pods again after the code thinks it's drained.  A different race.

So having written all that, I'm inclined to poll the operation.  You're right that seems cleaner.  But I still need a way to distinguish old node from new node, and I can't think of a better way than polling for age.  I guess I could poll for version with `kubectl get -o jsonpath='{.status.nodeInfo}' node/foo`.  That would work as long as we're sure that we're actually changing kubelet versions.

Comment (krousey) [inline cluster/gce/upgrade.sh]: Yes. I didn't mean not checking the nodes reported status. I just meant that checking the operation status is preferred to checking timestamps to see if the node rebooted. You still have all the other issues you mentioned.

Comment (mml) [inline cluster/gce/upgrade.sh]: Actually, there's another problem with the operation polling approach.  The recreate isn't a single operation.  It's two, apparently kicked off in succession asynchronously.  So I need to wait for the insert operation to be DONE.  But I need to make sure it's the *right* insert operation.  Consider the output below if I'm working on rx0c.  I don't want to accidentally match on the `insert` from `05-12` in this case if it's today.

This looks like I need more timestamp math.  WDYT?

```
% gcloud compute operations list --limit=6 --filter='targetLink:us-central1-f/instances/e2e-minion-group-*' --format='table[box](insertTime,startTime,endTime,operationType,targetLink,status)' --sort-by='~insertTime'                         
┌───────────────────────────────┬───────────────────────────────┬───────────────────────────────┬────────┬───────────────────────────────────────────────┬────────┐
│           TIMESTAMP           │           START_TIME          │            END_TIME           │  TYPE  │                     TARGET                    │ STATUS │
├───────────────────────────────┼───────────────────────────────┼───────────────────────────────┼────────┼───────────────────────────────────────────────┼────────┤
│ 2017-05-16T15:18:37.154-07:00 │ 2017-05-16T15:18:37.709-07:00 │ 2017-05-16T15:18:57.966-07:00 │ insert │ us-central1-f/instances/e2e-minion-group-rx0c │ DONE   │
│ 2017-05-16T15:17:26.260-07:00 │ 2017-05-16T15:17:26.599-07:00 │ 2017-05-16T15:18:35.020-07:00 │ delete │ us-central1-f/instances/e2e-minion-group-rx0c │ DONE   │
│ 2017-05-16T15:14:30.903-07:00 │ 2017-05-16T15:14:31.369-07:00 │ 2017-05-16T15:14:52.354-07:00 │ insert │ us-central1-f/instances/e2e-minion-group-6lw8 │ DONE   │
│ 2017-05-16T15:13:11.547-07:00 │ 2017-05-16T15:13:11.905-07:00 │ 2017-05-16T15:14:28.942-07:00 │ delete │ us-central1-f/instances/e2e-minion-group-6lw8 │ DONE   │
│ 2017-05-12T16:20:35.272-07:00 │ 2017-05-12T16:20:35.990-07:00 │ 2017-05-12T16:20:54.404-07:00 │ insert │ us-central1-f/instances/e2e-minion-group-rx0c │ DONE   │
│ 2017-05-12T16:19:26.596-07:00 │ 2017-05-12T16:19:26.908-07:00 │ 2017-05-12T16:20:33.207-07:00 │ delete │ us-central1-f/instances/e2e-minion-group-rx0c │ DONE   │
└───────────────────────────────┴───────────────────────────────┴───────────────────────────────┴────────┴───────────────────────────────────────────────┴────────┘
```

Comment (krousey) [inline cluster/gce/upgrade.sh]: So you don't think it would be sufficient to grab the latest insert op?

File: test/e2e/framework/service_util.go
Comment (thockin) [inline test/e2e/framework/service_util.go]: This is PROBABLY alright, but would we be safer to only set this for a testcase that is explicitly targetting this behavior?

File: cluster/gce/upgrade.sh
Comment (maisem) [inline cluster/gce/upgrade.sh]: You can watch for the `UID/externalID` of the node object to change. Or you can keep fetching the instance from GCE and wait for the ID to change. This is the authoritative way to know that the instance did indeed get recreated.
Once the ID changes, you need to wait for the node to enter `Ready` state.

`gcloud compute instances describe $INSTANCE --format='get(id)'`

Comment (mml) [inline cluster/gce/upgrade.sh]: @krousey, no because the "latest insert op" will actually be an old insert op for the first minute or so (maybe longer).

@maisem this is good.  I cache the id before I kick off `recreate-instances`, then poll until that id changes.  Then I poll `get nodes` until the `externalID` matches.  Then I can start polling node state.

```
% cluster/kubectl.sh get no e2e-minion-group-rx0c --output=jsonpath='{.spec.externalID}'
7990879616757056595
% gcloud compute instances describe e2e-minion-group-rx0c --format='get(id)'
7990879616757056595
```

File: test/e2e/framework/service_util.go
Comment (mml) [inline test/e2e/framework/service_util.go]: There is actually a hook in `RunOrFail` where I can add this logic (tweak function).  I'll externalize it and only set it up for this test.

File: cluster/gce/upgrade.sh
Comment (mml) [inline cluster/gce/upgrade.sh]: @maisem I think you're the best one to review this file.  Can you PTAL?

File: test/e2e/framework/service_util.go
Comment (mml) [inline test/e2e/framework/service_util.go]: Done.

File: test/e2e/framework/service_util.go
Comment (mml) [inline test/e2e/framework/service_util.go]: Done.  @krousey can you take a look at how I structured it?  I used the existing tweak hook, but the implementation is a little ugly.

**Comment by mml** [comment]
@k8s-bot gce etcd3 e2e test this

File: test/e2e/framework/service_util.go
Comment (krousey) [inline test/e2e/framework/service_util.go]: Minor nit, this tweak doesn't need to be attached to the jig. It could live along with the upgrade test that needs it.

File: test/e2e/framework/service_util.go
Comment (krousey) [inline test/e2e/framework/service_util.go]: Minor nit: Pod disruption budgets are orthogonal to services and don't really need to be attached to the jig. This could be created and managed completely separately from the service. It could be in its own utility or just along with the test.

File: test/e2e/framework/service_util.go
Comment (krousey) [inline test/e2e/framework/service_util.go]: So this still modifies every service test to now use a pod disruption budget? I'm not sure that's ok.

**Review by krousey — DISMISSED** [review]
Much better. I have a few minor nits that you can disregard if you disagree. The only remaining question I have is if it's ok for every service test to create and wait on a PDB now.

File: cluster/gce/upgrade.sh
Comment (mml) [inline cluster/gce/upgrade.sh]: I could assign the output of `list-instances` to a variable and then run a for loop.  Main advantage is that `gcloud` won't block writing to a full pipe and end up freaking out because it runs so long, which could happen in a large cluster.

Comment (mml) [inline cluster/gce/upgrade.sh]: Done.

File: test/e2e/framework/service_util.go
Comment (mml) [inline test/e2e/framework/service_util.go]: It needs access to the jig to get the correct labels.  I could pass the jig as an argument, or the test could grab the labels from the jig and pass them as an argument.  In the end I left it here because the tweak requires requires a bunch of internal knowledge about all the rest of the stuff the jig does.

**Comment by mml** [comment]
> Much better. I have a few minor nits that you can disregard if you disagree. The only remaining question I have is if it's ok for every service test to create and wait on a PDB now.

No.  The waiting step will introduce pointless flakiness.  I'll move that out.

**Comment by mml** [comment]
cc @monopole

File: test/e2e/upgrades/services.go
Comment (maisem) [inline test/e2e/upgrades/services.go]: This should be enabled for GKE as well.

**Comment by maisem** [comment]
@mml - upgrade.sh changes LGTM.

File: test/e2e/upgrades/services.go
Comment (maisem) [inline test/e2e/upgrades/services.go]: GCE/GKE

**Review by maisem — APPROVED** [review]
LGTM after comment fix.

File: test/e2e/upgrades/services.go
Comment (mml) [inline test/e2e/upgrades/services.go]: Done.

**Comment by lavalamp** [comment]
/approve since @krousey is on vacation

(although someone else should review the shell :) )

**Comment by maisem** [comment]
@k8s-bot unit test this

**Comment by fejta** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref: #46827

**Comment by shashidharatd** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref: #46827

**Comment by mml** [comment]
@shashidharatd thanks!

**Comment by mml** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this #38275

**Comment by mml** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Pull request merged**
